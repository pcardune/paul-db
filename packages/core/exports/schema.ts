export { ColumnTypes as type } from "../schema/columns/ColumnType.ts"
export { column, computedColumn } from "../schema/columns/ColumnBuilder.ts"

import { TableSchema } from "../schema/schema.ts"
export const table = TableSchema.create
export { create as db } from "../schema/DBSchema.ts"

export type {
  InsertRecordForTableSchema as InferInsertRecord,
  StoredRecordForTableSchema as InferRecord,
} from "../schema/schema.ts"
export type { DBModel } from "../db/DbFile.ts"

export type { HeapFileTableInfer as InferTable } from "../tables/TableStorage.ts"
