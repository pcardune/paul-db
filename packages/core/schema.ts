// deno-lint-ignore-file no-explicit-any
import { Comparator, EqualityChecker } from "./DiskBTree.ts"

export const ColumnTypes = {
  any<T>() {
    return new ColumnType<T>({ isValid: (_value: T): _value is T => true })
  },
  positiveNumber() {
    return new ColumnType<number>({
      isValid: (value) => value > 0,
    })
  },
}

export class ColumnType<T> {
  isValid: (value: T) => boolean
  isEqual: EqualityChecker<T>
  compare: Comparator<T>
  constructor({
    isValid,
    equals = (a: T, b: T) => a === b,
    compare = (a: T, b: T) => (a > b ? 1 : a < b ? -1 : 0),
  }: {
    isValid: (value: T) => boolean
    equals?: EqualityChecker<T>
    compare?: Comparator<T>
  }) {
    this.isValid = isValid
    this.isEqual = equals
    this.compare = compare
  }
}

export class ColumnSchema<
  Name extends string,
  ValueT,
  UniqueT extends boolean,
> {
  readonly name: Name
  readonly unique: UniqueT
  readonly type: ColumnType<ValueT>

  private constructor(
    name: Name,
    type: ColumnType<ValueT>,
    unique: UniqueT,
  ) {
    this.name = name
    this.type = type
    this.unique = unique
  }

  static create<
    Name extends string,
    ValueT,
  >(
    name: Name,
    type: ColumnType<ValueT>,
  ) {
    return new ColumnSchema(name, type, false)
  }

  withName<NewName extends string>(name: NewName) {
    return new ColumnSchema<NewName, ValueT, UniqueT>(
      name,
      this.type,
      this.unique,
    )
  }

  makeUnique(): ColumnSchema<Name, ValueT, true> {
    return new ColumnSchema(this.name, this.type, true)
  }
}

export type SomeColumnSchema = ColumnSchema<string, any, boolean>

type PushTuple<T extends any[], V> = [...T, V]

export type RecordForColumnSchema<CS extends SomeColumnSchema[]> = {
  [K in CS[number]["name"]]:
    Extract<CS[number], { name: K }>["type"]["isValid"] extends (
      value: infer V,
    ) => boolean ? V
      : never
}

export type RecordForTableSchema<TS extends TableSchema<any, any>> =
  RecordForColumnSchema<
    TS["columns"]
  >

export class TableSchema<
  TableName extends string,
  ColumnSchemasT extends SomeColumnSchema[],
> {
  name: TableName
  columns: ColumnSchemasT

  private constructor(name: TableName, columns: ColumnSchemasT) {
    this.name = name
    this.columns = columns
  }

  getColumns(): ColumnSchemasT {
    return this.columns
  }

  static create<Name extends string>(name: Name) {
    return new TableSchema(name, [])
  }

  isValidRecord(record: RecordForColumnSchema<ColumnSchemasT>): boolean {
    for (const column of this.columns) {
      const value = record[column.name as keyof typeof record]
      if (!column.type.isValid(value)) {
        return false
      }
    }
    return true
  }

  withColumn<CName extends string, CValue, CUnique extends boolean>(
    column: ColumnSchema<CName, CValue, CUnique>,
  ): TableSchema<
    TableName,
    PushTuple<ColumnSchemasT, ColumnSchema<CName, CValue, CUnique>>
  >
  withColumn<CName extends string, CValue>(
    name: CName,
    type: ColumnType<CValue>,
  ): TableSchema<
    TableName,
    PushTuple<ColumnSchemasT, ColumnSchema<CName, CValue, false>>
  >
  withColumn<CName extends string, CValue>(
    name: CName,
    type: ColumnType<CValue>,
    options: {
      unique: true
    },
  ): TableSchema<
    TableName,
    PushTuple<ColumnSchemasT, ColumnSchema<CName, CValue, true>>
  >
  withColumn<
    CName extends string,
    CValue,
    CUnique extends boolean,
  >(
    nameOrColumn: CName | ColumnSchema<CName, CValue, CUnique>,
    type?: ColumnType<CValue>,
    options?: { unique: true },
  ): TableSchema<
    TableName,
    PushTuple<ColumnSchemasT, ColumnSchema<CName, CValue, CUnique>>
  > {
    if (typeof nameOrColumn === "string") {
      if (type) {
        if (options) {
          return new TableSchema(this.name, [
            ...this.columns,
            ColumnSchema.create(nameOrColumn, type)
              .makeUnique() as ColumnSchema<CName, CValue, CUnique>,
          ])
        }
        return new TableSchema(this.name, [
          ...this.columns,
          ColumnSchema.create(nameOrColumn, type) as ColumnSchema<
            CName,
            CValue,
            CUnique
          >,
        ])
      } else {
        throw new Error("Type and options are required")
      }
    }
    return new TableSchema(this.name, [
      ...this.columns,
      nameOrColumn,
    ])
  }
}
