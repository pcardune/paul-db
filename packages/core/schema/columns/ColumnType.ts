import { IStruct, Struct } from "../../binary/Struct.ts"
import { Comparator, EqualityChecker } from "../../types.ts"
import * as ColumnTypes from "./types.ts"

export { ColumnTypes }

/**
 * Parses a string into a ColumnType instance
 *
 * @throws Error if the type is unknown
 */
export function getColumnTypeFromString(
  type: `${string}[]`,
): ColumnType<unknown[]>
export function getColumnTypeFromString(
  type: `${string}?`,
): ColumnType<unknown | null>
export function getColumnTypeFromString(type: string): ColumnType<unknown>
export function getColumnTypeFromString(
  type: `${string}[]` | string,
): ColumnType<unknown> | ColumnType<unknown[]> {
  const t = type.toLowerCase()

  if (t.endsWith("?")) {
    return getColumnTypeFromString(t.slice(0, -1)).nullable()
  }
  if (t.endsWith("[]")) {
    return getColumnTypeFromString(t.slice(0, -2)).array()
  }

  if (t in ColumnTypes) {
    return (ColumnTypes as Record<string, () => ColumnType<unknown>>)[t]()
  }
  if (t === "serial") {
    return new SerialUInt32ColumnType() as ColumnType<unknown>
  }
  throw new Error(`Unknown type: ${type}`)
}

/**
 * Converts a SQL type to a ColumnType instance
 */
export function getColumnTypeFromSQLType(sqlType: string): ColumnType<any> {
  if (sqlType.endsWith("[]")) {
    return getColumnTypeFromSQLType(sqlType.slice(0, -2)).array()
  }
  switch (sqlType) {
    case "TEXT":
    case "VARCHAR":
    case "CHAR":
      return ColumnTypes.string()
    case "SMALLINT":
      return ColumnTypes.int16()
    case "INT":
    case "INTEGER":
      return ColumnTypes.int32()
    case "FLOAT":
    case "REAL":
    case "DOUBLE":
      return ColumnTypes.float()
    case "BOOLEAN":
      return ColumnTypes.boolean()
    case "UUID":
      return ColumnTypes.uuid()
    case "JSON":
    case "JSONB":
      return ColumnTypes.json()
    case "DATE":
      return ColumnTypes.date()
    case "TIMESTAMP":
      return ColumnTypes.timestamp()
    case "BLOB":
      return ColumnTypes.blob()
    case "SERIAL":
      return new SerialUInt32ColumnType()
    default:
      throw new Error(`Unknown SQL type: ${sqlType}`)
  }
}

/**
 * Represents a column type in a database table
 */
export class ColumnType<T = unknown> {
  readonly name: string
  isValid: (value: T) => boolean
  isEqual: EqualityChecker<T>
  compare: Comparator<T>
  serializer?: IStruct<T>

  constructor({
    name,
    isValid,
    equals = (a: T, b: T) => a === b,
    compare = (a: T, b: T) => (a > b ? 1 : a < b ? -1 : 0),
    serializer,
  }: {
    name: string
    isValid: (value: T) => boolean
    equals?: EqualityChecker<T>
    compare?: Comparator<T>
    serializer?: IStruct<T>
  }) {
    this.name = name
    this.isValid = isValid
    this.isEqual = equals
    this.compare = compare
    this.serializer = serializer
  }

  /**
   * Takes a ColumnType and returns a new ColumnType that allows arrays of values
   */
  array(): ColumnType<T[]> {
    return new ColumnType<T[]>({
      name: this.name + "[]",
      isValid: (value) => Array.isArray(value) && value.every(this.isValid),
      equals: (a, b) =>
        a.length === b.length && a.every((v, i) => this.isEqual(v, b[i])),
      compare: (a, b) => {
        for (let i = 0; i < Math.min(a.length, b.length); i++) {
          const cmp = this.compare(a[i], b[i])
          if (cmp !== 0) return cmp
        }
        return a.length - b.length
      },
      serializer: this.serializer?.array(),
    })
  }

  /**
   * Takes a ColumnType and returns a new ColumnType that allows null values
   */
  nullable(): ColumnType<T | null> {
    return new ColumnType<T | null>({
      name: this.name + "?",
      isValid: (value) => value === null || this.isValid(value),
      equals: (a, b) =>
        a === b || (a !== null && b !== null && this.isEqual(a, b)),
      compare: (a, b) => {
        if (a === b) return 0
        if (a === null) return -1
        if (b === null) return 1
        return this.compare(a, b)
      },
      serializer: this.serializer?.nullable(),
    })
  }
}

export class SerialUInt32ColumnType extends ColumnType<number> {
  constructor() {
    super({
      name: "serial",
      isValid: (value) => value >= 0,
      serializer: Struct.uint32,
    })
  }
}
