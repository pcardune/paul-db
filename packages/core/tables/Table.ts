// deno-lint-ignore-file no-explicit-any
import { Index } from "../indexes/Index.ts"
import {
  ColumnIndexConfig,
  InputForComputedColumnSchema,
  InsertRecordForTableSchema,
  OutputForComputedColumnSchema,
  SomeColumnSchema,
  SomeComputedColumnSchema,
  StoredRecordForTableSchema,
  TableSchema,
  ValueForColumnSchema,
} from "../schema/schema.ts"
import { ITableStorage } from "./TableStorage.ts"
import { FilterTuple } from "../typetools.ts"
import { INodeId } from "../indexes/BTreeNode.ts"
import { AsyncIterableWrapper } from "../async.ts"

/**
 * A helper type that lets you declare a table type from a given
 * schema and storage type.
 */
export type TableInfer<SchemaT, StorageT> = SchemaT extends
  TableSchema<infer TName, infer ColumnSchemasT, infer ComputedColumnSchemasT>
  ? StorageT extends
    ITableStorage<infer RowIdT, StoredRecordForTableSchema<SchemaT>> ? Table<
      RowIdT,
      TName,
      ColumnSchemasT,
      ComputedColumnSchemasT,
      SchemaT,
      ITableStorage<RowIdT, StoredRecordForTableSchema<SchemaT>>
    >
  : never
  : never

export class Table<
  RowIdT,
  TName extends string,
  ColumnSchemasT extends SomeColumnSchema[],
  ComputedColumnSchemasT extends SomeComputedColumnSchema[],
  SchemaT extends TableSchema<TName, ColumnSchemasT, ComputedColumnSchemasT>,
  StorageT extends ITableStorage<RowIdT, StoredRecordForTableSchema<SchemaT>>,
> {
  private schema: SchemaT
  private data: StorageT
  private _allIndexes: Map<string, Index<unknown, RowIdT, INodeId>>

  constructor(init: {
    schema: SchemaT
    data: StorageT
    indexes: Map<string, Index<unknown, RowIdT, INodeId>>
  }) {
    this.schema = init.schema
    this.data = init.data

    this._allIndexes = init.indexes
  }

  async insertManyAndReturn(
    records: InsertRecordForTableSchema<SchemaT>[],
  ): Promise<StoredRecordForTableSchema<SchemaT>[]> {
    const rows: StoredRecordForTableSchema<SchemaT>[] = []
    for (const record of records) {
      rows.push(await this.insertAndReturn(record))
    }
    return rows
  }

  async insertMany(
    records: InsertRecordForTableSchema<SchemaT>[],
  ): Promise<RowIdT[]> {
    const rowIds: RowIdT[] = []
    for (const record of records) {
      rowIds.push(await this.insert(record))
    }
    return rowIds
  }

  async insertAndReturn(
    record: InsertRecordForTableSchema<SchemaT>,
  ): Promise<StoredRecordForTableSchema<SchemaT>> {
    const id = await this.insert(record)
    return this.data.get(id) as Promise<StoredRecordForTableSchema<SchemaT>>
  }

  async insert(record: InsertRecordForTableSchema<SchemaT>): Promise<RowIdT> {
    const validation = this.schema.isValidInsertRecord(record)
    if (!validation.valid) {
      throw new Error("Invalid record: " + validation.reason)
    }

    for (const column of this.schema.columns) {
      if ((record as any)[column.name] == null && column.defaultValueFactory) {
        const value = column.defaultValueFactory()
        if (!column.type.isValid(value)) {
          throw new Error(`Default value for ${column.name} is invalid`)
        }
        ;(record as any)[column.name] = value
      }
    }

    for (const column of this.schema.columns) {
      if (!column.unique) continue
      const index = this._allIndexes.get(column.name)
      if (!index) {
        throw new Error(
          `Column ${column.name} is not indexed but is marked as unique`,
        )
      }
      const value = (record as any)[column.name]
      if (await index.has(value)) {
        throw new Error(
          `Record with given ${column.name} value already exists`,
        )
      }
    }
    for (const column of this.schema.computedColumns) {
      if (!column.unique) continue
      const index = this._allIndexes.get(column.name)
      if (!index) {
        throw new Error(
          `Column ${column.name} is not indexed but is marked as unique`,
        )
      }
      const value = column.compute(record as any)
      if (await index.has(value)) {
        throw new Error(
          `Record with given ${column.name} value already exists`,
        )
      }
    }

    const id = await this.data.insert(record)
    for (const column of this.schema.columns) {
      const index = this._allIndexes.get(column.name)
      if (index) {
        await index.insert((record as any)[column.name], id)
      } else if (column.indexed) {
        throw new Error(`Column ${column.name} is not indexed`)
      }
    }
    for (const column of this.schema.computedColumns) {
      const index = this._allIndexes.get(column.name)
      if (index) {
        await index.insert(column.compute(record), id)
      } else if (column.indexed) {
        throw new Error(`Column ${column.name} is not indexed`)
      }
    }
    await this.data.commit()
    return id
  }

  get(id: RowIdT): Promise<StoredRecordForTableSchema<SchemaT> | undefined> {
    return this.data.get(id)
  }

  async remove(id: RowIdT): Promise<void> {
    await this.data.remove(id)
    await this.data.commit()
  }

  async lookupUniqueOrThrow<
    IName extends FilterTuple<
      SchemaT["columns"] | SchemaT["computedColumns"],
      { unique: true }
    >["name"],
    ValueT extends
      | ValueForColumnSchema<
        FilterTuple<
          SchemaT["columns"],
          { name: IName }
        >
      >
      | InputForComputedColumnSchema<
        FilterTuple<SchemaT["computedColumns"], { name: IName }>
      >,
  >(
    indexName: IName,
    value: ValueT,
  ): Promise<Readonly<StoredRecordForTableSchema<SchemaT>>> {
    const result = await this.lookupUnique(indexName, value)
    if (!result) {
      console.error(`Record not found for`, indexName, value)
      throw new Error(`Record not found`)
    }
    return result
  }

  async lookupUnique<
    IName extends FilterTuple<
      SchemaT["columns"] | SchemaT["computedColumns"],
      { unique: true }
    >["name"],
    ValueT extends
      | ValueForColumnSchema<
        FilterTuple<
          SchemaT["columns"],
          { name: IName }
        >
      >
      | InputForComputedColumnSchema<
        FilterTuple<SchemaT["computedColumns"], { name: IName }>
      >,
  >(
    indexName: IName,
    value: ValueT,
  ): Promise<Readonly<StoredRecordForTableSchema<SchemaT>> | undefined> {
    const index = this._allIndexes.get(indexName)
    if (!index) {
      throw new Error(`Index ${indexName} does not exist`)
    }
    const computedColumn = this.schema.computedColumns.find((c) =>
      c.name === indexName
    )
    let valueToLookup = value
    if (computedColumn != null) {
      valueToLookup = computedColumn.compute(value)
    }
    const rowIds = await index.get(valueToLookup)
    if (rowIds.length === 0) {
      return
    }
    return this.data.get(rowIds[0])
  }

  async lookup<
    IName extends FilterTuple<
      SchemaT["columns"],
      { indexed: ColumnIndexConfig }
    >["name"],
    ValueT extends ValueForColumnSchema<
      FilterTuple<SchemaT["columns"], { name: IName }>
    >,
  >(
    indexName: IName,
    value: ValueT,
  ): Promise<Readonly<StoredRecordForTableSchema<SchemaT>>[]> {
    const index = this._allIndexes.get(indexName)
    if (!index) {
      throw new Error(`Index ${indexName} does not exist`)
    }
    const computedColumn = this.schema.computedColumns.find((c) =>
      c.name === indexName
    )
    let valueToLookup = value
    if (computedColumn != null) {
      valueToLookup = computedColumn.compute(value)
    }
    return Promise.all(
      (await index.get(valueToLookup)).map((id) => {
        return this.data.get(id) as Promise<StoredRecordForTableSchema<SchemaT>>
      }),
    )
  }

  async lookupComputed<
    IName extends FilterTuple<
      SchemaT["computedColumns"],
      { indexed: ColumnIndexConfig }
    >["name"],
    ValueT extends OutputForComputedColumnSchema<
      FilterTuple<SchemaT["computedColumns"], { name: IName }>
    >,
  >(indexName: IName, value: ValueT) {
    const index = this._allIndexes.get(indexName)
    if (!index) {
      throw new Error(`Index ${indexName} does not exist`)
    }
    const column = this.schema.computedColumns.find((c) => c.name === indexName)
    if (!column) {
      throw new Error(`Column ${indexName} is not a computed column`)
    }
    return Promise.all(
      (await index.get(value)).map((id) =>
        this.data.get(id) as Promise<StoredRecordForTableSchema<SchemaT>>
      ),
    )
  }

  iterate(): AsyncIterableWrapper<
    StoredRecordForTableSchema<SchemaT>
  > {
    const pkIndex = this._allIndexes.get(this.schema.columns[0].name)
    if (!pkIndex) {
      throw new Error("Primary key index not found")
    }

    const data = this.data

    const iterator = {
      [Symbol.asyncIterator]: async function* () {
        const allKeys = await pkIndex.getRange({})
        for (const key of allKeys) {
          for (const id of key.vals) {
            const record = await data.get(id)
            if (record) {
              yield record
            }
          }
        }
      },
    }
    return new AsyncIterableWrapper(iterator)
  }

  scanIter<
    IName extends ColumnSchemasT[number]["name"],
    IValue extends StoredRecordForTableSchema<SchemaT>[IName],
  >(
    columnName: IName,
    value: IValue,
  ): AsyncIterableWrapper<StoredRecordForTableSchema<SchemaT>> {
    const columnType =
      this.schema.columns.find((c) => c.name === columnName)!.type
    return this.iterate().filter((record) =>
      columnType.isEqual(record[columnName], value)
    )
  }

  scan<
    IName extends ColumnSchemasT[number]["name"],
    IValue extends StoredRecordForTableSchema<SchemaT>[IName],
  >(
    columnName: IName,
    value: IValue,
  ): Promise<StoredRecordForTableSchema<SchemaT>[]> {
    return this.scanIter(columnName, value).toArray()
  }
}
