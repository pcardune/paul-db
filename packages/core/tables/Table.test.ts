import { beforeAll, describe, it } from "jsr:@std/testing/bdd"
import { expect } from "jsr:@std/expect"
import { assertSnapshot } from "jsr:@std/testing/snapshot"
import { Table, TableInfer } from "./Table.ts"
import { StoredRecordForTableSchema, TableSchema } from "../schema/schema.ts"
import {
  HeapFileRowId,
  InMemoryTableStorage,
  JsonFileTableStorage,
} from "./TableStorage.ts"
import { ColumnType, ColumnTypes } from "../schema/ColumnType.ts"
import { DbFile } from "../db/DbFile.ts"
import { generateTestFilePath } from "../testing.ts"
import { column, computedColumn } from "../schema/ColumnSchema.ts"

const peopleSchema = TableSchema.create("people")
  .with(column("name", ColumnTypes.any<string>()))
  .with(column("age", ColumnTypes.positiveNumber()))

describe("Create, Read, and Delete", () => {
  it("lets you insert and retrieve records", async () => {
    const people = new Table(await InMemoryTableStorage.forSchema(peopleSchema))

    const aliceId = await people.insert({ name: "Alice", age: 12 })
    const bobId = await people.insert({ name: "Bob", age: 12 })
    expect(await people.get(aliceId)).toEqual({
      name: "Alice",
      age: 12,
    })
    expect(await people.get(bobId)).toEqual({
      name: "Bob",
      age: 12,
    })
  })
  it("You can delete records", async () => {
    const people = new Table(
      await InMemoryTableStorage.forSchema(peopleSchema),
    )
    const aliceId = await people.insert({ name: "Alice", age: 12 })
    const bobId = await people.insert({ name: "Bob", age: 12 })
    people.remove(aliceId)
    expect(await people.get(aliceId)).toBeUndefined()
    expect(await people.get(bobId)).toEqual({
      name: "Bob",
      age: 12,
    })
  })
})

describe("Insert Validation", () => {
  it("should not allow you to insert records with invalid schema", async () => {
    const people = new Table(
      await InMemoryTableStorage.forSchema(peopleSchema.with(
        column("uuid", ColumnTypes.uuid()).defaultTo(() => crypto.randomUUID()),
      )),
    )
    await people.insert({ name: "Alice", age: 12 })
    expect(people.insert({ name: "Alice", age: -12 })).rejects.toThrow(
      "Invalid record: Invalid value for column age",
    )
    expect(
      people.insert({
        name: "Alice",
        age: 12,
        uuid: "not-valid-uuid-dispite-typecheck",
      }),
    ).rejects.toThrow(
      "Invalid record: Invalid value for column uuid",
    )
  })

  it("Throws when inserting values that don't satisfy the column type", async () => {
    const schema = peopleSchema
      .with(
        column(
          "favoriteOdd",
          new ColumnType({
            name: "oddNumber",
            isValid: (value: number) => value % 2 === 1,
          }),
        ),
      )
    const oddPeople = new Table(await InMemoryTableStorage.forSchema(schema))

    oddPeople.insert({ name: "Alice", age: 13, favoriteOdd: 13 })
    expect(oddPeople.insert({ name: "Alice", age: 12, favoriteOdd: 12 }))
      .rejects.toThrow("Invalid record")
  })
})

const phoneNumberType = new ColumnType<string>({
  name: "phoneNumber",
  isValid: (value) => /^[\d-]+$/.test(value),
  equals: (a, b) => a.replace(/-/g, "") === b.replace(/-/g, ""),
  compare: (a, b) => a.replace(/-/g, "").localeCompare(b.replace(/-/g, "")),
})

describe("Uniqueness Constraints", () => {
  it("enforces uniqueness constraints", async () => {
    const schema = peopleSchema
      .with(column("ssn", ColumnTypes.any<string>()).unique())
    const people = new Table(await InMemoryTableStorage.forSchema(schema))

    await people.insert({ name: "Alice", age: 12, ssn: "123-45-6789" })
    expect(people.insert({ name: "Alice", age: 12, ssn: "123-45-6789" }))
      .rejects.toThrow("Record with given ssn value already exists")
    await people.insert({ name: "Alice", age: 12, ssn: "123-45-6-7-89" })
  })

  it("utilizes column type to determine uniqueness", async () => {
    const schema = peopleSchema
      .with(column("phone", phoneNumberType).unique())
    const people = new Table(await InMemoryTableStorage.forSchema(schema))
    await people.insert({ name: "Alice", age: 12, phone: "123-867-5309" })
    expect(
      people.insert({ name: "Alice", age: 12, phone: "123-8675-309" }),
    ).rejects.toThrow("Record with given phone value already exists")
  })
})

describe("Querying", () => {
  describe("Table.iterate()", () => {
    it("lets you iterate over the entire contents of the table", async () => {
      const people = new Table(
        await InMemoryTableStorage.forSchema(peopleSchema),
      )
      await people.insertMany([
        { name: "Alice", age: 30 },
        { name: "Bob", age: 30 },
        { name: "Charlie", age: 35 },
      ])

      const data = await people.iterate()
        .filter((r) => r.name.toLowerCase().includes("a"))
        .toArray()

      expect(data).toEqual([
        { name: "Alice", age: 30 },
        { name: "Charlie", age: 35 },
      ])
    })
  })

  describe("Table.scan()", () => {
    it("can query records by scanning the entire table", async () => {
      const people = new Table(
        await InMemoryTableStorage.forSchema(peopleSchema),
      )
      await people.insertMany([
        { name: "Alice", age: 30 },
        { name: "Bob", age: 30 },
        { name: "Charlie", age: 35 },
      ])

      expect(await people.scan("age", 30)).toEqual([
        { name: "Alice", age: 30 },
        { name: "Bob", age: 30 },
      ])
    })

    it("uses the underlying column type for equality testing", async () => {
      const peopleSchema = TableSchema.create("people").with(
        column("name", ColumnTypes.any<string>()),
      ).with(column("email", ColumnTypes.caseInsensitiveString()))
      const people = new Table(
        await InMemoryTableStorage.forSchema(peopleSchema),
      )
      await people.insertMany([
        { name: "Alice", email: "alice@example.com" },
        { name: "Alice 2", email: "Alice@example.com" },
        { name: "Bob", email: "bob@example.com" },
        { name: "Charlie", email: "charlie@website.com" },
      ])
      expect(await people.scan("email", "ALICE@EXAMPLE.COM")).toEqual([
        { name: "Alice", email: "alice@example.com" },
        { name: "Alice 2", email: "Alice@example.com" },
      ])
    })
  })

  describe("Table.lookup()", () => {
    const indexedPeopleSchema = TableSchema.create("people")
      .with(column("name", ColumnTypes.any<string>()).index())
      .with(column("phone", phoneNumberType))
      .with(column("age", ColumnTypes.positiveNumber()).index())
      .withComputedColumn(
        computedColumn(
          "lowerCaseName",
          ColumnTypes.string(),
          (input: { name: string }) => input.name.toLowerCase(),
        ).index(),
      )
    let people: TableInfer<
      typeof indexedPeopleSchema,
      InMemoryTableStorage<
        number,
        StoredRecordForTableSchema<typeof indexedPeopleSchema>
      >
    >
    beforeAll(async () => {
      people = new Table(
        await InMemoryTableStorage.forSchema(indexedPeopleSchema),
      )
    })

    beforeAll(async () => {
      await people.insertMany([
        { name: "Alice", age: 25, phone: "123-456-7890" },
        { name: "Bob", age: 35, phone: "123-456-7891" },
        { name: "Charlie", age: 25, phone: "123-456-7892" },
      ])
    })

    it("lets you query using an index", async () => {
      expect(await people.lookup("age", 25)).toEqual([
        { name: "Alice", age: 25, phone: "123-456-7890" },
        { name: "Charlie", age: 25, phone: "123-456-7892" },
      ])
      expect(await people.lookup("age", 35)).toEqual([
        { name: "Bob", age: 35, phone: "123-456-7891" },
      ])
    })

    it("lets you query using an index on a computed column", async () => {
      expect(await people.lookupComputed("lowerCaseName", "alice"))
        .toEqual([
          { name: "Alice", age: 25, phone: "123-456-7890" },
        ])
    })

    it.skip("will throw if you lookup a computed indexed column with an invalid value", async () => {
      expect(await people.lookupComputed("lowerCaseName", "ALICE")).rejects
        .toThrow(
          "Invalid value for column lowerCaseName",
        )
    })
  })
})

Deno.test({
  name: "json table storage",
  permissions: { read: true, write: true },
  fn: async (t) => {
    using tempFile = generateTestFilePath("people.json")
    const storage = await JsonFileTableStorage.forSchema(
      peopleSchema,
      tempFile.filePath,
    )
    const people = new Table(storage)
    await people.insert({ name: "Alice", age: 12 })
    await people.insert({ name: "Bob", age: 12 })
    const f = Deno.readTextFileSync(tempFile.filePath)
    await assertSnapshot(t, f)
  },
})

Deno.test("HeapFileTableStorage", async (t) => {
  using tempFile = generateTestFilePath("people.data")

  const schema = TableSchema.create("people")
    .with(column("id", ColumnTypes.string()).unique())
    .with(column("firstName", ColumnTypes.string()))
    .with(column("lastName", ColumnTypes.string()))
    .with(column("age", ColumnTypes.uint32()))
    .with(column("likesIceCream", ColumnTypes.boolean()))

  async function useTableResources(
    filePath: string,
    { truncate = false } = {},
  ) {
    const dbFile = await DbFile.open(filePath, { create: true, truncate })
    const table = new Table(await dbFile.getTableStorage(schema))
    return { table, dbFile, [Symbol.dispose]: () => dbFile.close() }
  }

  using resources = await useTableResources(tempFile.filePath, {
    truncate: true,
  })
  const { table: people } = resources
  let aliceRowId: HeapFileRowId
  const otherRowIds: HeapFileRowId[] = []

  await t.step(".insert()", async () => {
    aliceRowId = await people.insert({
      id: "1",
      firstName: "Alice",
      lastName: "Jones",
      age: 25,
      likesIceCream: true,
    })
    for (let i = 2; i <= 100; i++) {
      const rowId = await people.insert({
        id: i.toString(),
        firstName: `Person ${i}`,
        lastName: `Lastname ${i}`,
        age: i,
        likesIceCream: i % 2 === 0,
      })
      otherRowIds.push(rowId)
    }
  })

  await t.step(".set()", async () => {
    const alice = await people.get(aliceRowId)
    let newRowId = await people.set(aliceRowId, { ...alice!, age: 26 })

    // the row id should stay the same because the record fits into
    // the existing slot
    expect(newRowId).toEqual(aliceRowId)
    expect(await people.lookupUnique("id", "1")).toEqual({
      id: "1",
      firstName: "Alice",
      lastName: "Jones",
      age: 26,
      likesIceCream: true,
    })

    // the row id should change because the record no longer fits into
    // the existing slot
    newRowId = await people.set(aliceRowId, {
      ...alice!,
      age: 27,
      lastName: "from the books about wonderland",
    })
    expect(newRowId).not.toMatchObject(aliceRowId)
    expect(await people.get(aliceRowId), "old row id will still work").toEqual({
      id: "1",
      firstName: "Alice",
      lastName: "from the books about wonderland",
      age: 27,
      likesIceCream: true,
    })
    expect(
      await people.lookupUnique("id", "1"),
      "looking up from an index will work",
    ).toEqual({
      id: "1",
      firstName: "Alice",
      lastName: "from the books about wonderland",
      age: 27,
      likesIceCream: true,
    })

    // we can still find the next record that was inserted
    expect(await people.get(otherRowIds[0])).toEqual({
      age: 2,
      firstName: "Person 2",
      id: "2",
      lastName: "Lastname 2",
      likesIceCream: true,
    })
  })

  using resources2 = await useTableResources(tempFile.filePath)
  const fixtures = [
    { resources, name: "table that was written to" },
    { resources: resources2, name: "a just opened file" },
  ]
  for (const { resources: { table: people }, name } of fixtures) {
    await t.step(`reads [${name}]`, async (t) => {
      await t.step(
        "Table.get()",
        async () => {
          expect(await people.get(aliceRowId)).toEqual({
            id: "1",
            firstName: "Alice",
            lastName: "from the books about wonderland",
            age: 27,
            likesIceCream: true,
          })
        },
      )

      await t.step(
        "Table.lookupUnique()",
        async () => {
          expect(await people.lookupUnique("id", "1")).toEqual({
            id: "1",
            firstName: "Alice",
            lastName: "from the books about wonderland",
            age: 27,
            likesIceCream: true,
          })
        },
      )
    })
  }
})
