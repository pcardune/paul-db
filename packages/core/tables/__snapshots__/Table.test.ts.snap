export const snapshot = {};

snapshot[`json table storage 1`] = `
'{
  "1": {
    "id": "1",
    "name": "Alice",
    "age": 12
  },
  "2": {
    "id": "2",
    "name": "Bob",
    "age": 12
  }
}'
`;
